### ISSUE TITLE
Many "undefined reference" errors when linking to qt5-static

### ISSUE THREAD

**Issue by Morilli** [msys2/MINGW-packages#8537]
I'm trying to compile and link the project https://github.com/moonshadow565/lolcustomskin-tools using cmake and the mingw qt5-static package. When using the normal qt5 package, I can install the project fine, but using the qt5-static package, I get 2.5k linker errors on the final link step. I have even tried compiling qt5 from source which absolutely did suck because qt5 isn't made to be compiled, but to no help.
I really am at a lost here to what to try now, I just want to get this to work.

Here's all the linker errors: 
[log.txt](https://github.com/msys2/MINGW-packages/files/6411749/log.txt)

**Comment by Biswa96** [comment]
It looks like you are using ucrt64 toolchain. Have you tried with mingw64 toolchain (the blue icon in start menu)?

**Comment by Morilli** [comment]
Yes, I have installed with mingw64 toolchain before. The qt5-static package is also only available on the (non-ucrt) mingw64 repo, so this log I posted is only from my self-compiled ucrt attempt, the previous one was normal mingw64 toolchain only with preinstalled qt5-static package.

**Comment by revelator** [comment]
most of the problem seems to stem from the fact that the static qt5 only accepts static library dependencies :S also you need to link to a good deal more runtime libraries in static builds as the compiler has no notion on which msvc or ucrt runtimes the source might depend on.  dll builds are somewhat easier as the dependant dll's just call the nessesary runtimes dynamically.

**Comment by Morilli** [comment]
There should be no additional library dependencies, neither static nor dynamic, so there should be no such issues. As a note msvc builds the project perfectly fine.

In case that helps here is a more or less minimal reproduction example that showcases the fact that mingw's static qt builds fail to link, while it builds just fine under msvc:
[qtbullshit.zip](https://github.com/msys2/MINGW-packages/files/6414388/qtbullshit.zip)

**Comment by 1480c1** [comment]
qt5 fails to build with a lot of things as the pkg-config files don't have all the little qt libs, and they only declare them in their prl source files https://code.videolan.org/videolan/vlc/blob/master/contrib/src/qt/AddStaticLink.sh this is the script vlc uses to fix the pkg-config files and it can be shortened to
```bash
_add_static_link() {
    local PRL_SOURCE=$LOCALDESTDIR/$2/lib$3.prl LIBS
    [[ -f $PRL_SOURCE ]] || PRL_SOURCE=$LOCALDESTDIR/$2/$3.prl
    [[ ! -f $PRL_SOURCE ]] && return 1
    LIBS=$(sed -e "
        /QMAKE_PRL_LIBS =/ {
            s@QMAKE_PRL_LIBS =@@
            s@$LOCALDESTDIR/lib@\${libdir}@g
            s@\$\$\[QT_INSTALL_LIBS\]@\${libdir}@g
            p
        }
        d" "$PRL_SOURCE" | grep -v QMAKE_PRL_LIBS_FOR_CMAKE)
    sed -i.bak "
        s# -l$1# -l$3 -l$1#
        s#Libs.private:.*#& $LIBS -L\${prefix}/$2#
        " "$LOCALDESTDIR/lib/pkgconfig/$1.pc"
}
```
and used like
```bash
_add_static_link Qt5Gui plugins/platforms qwindows
```

I would recommend if you are going to go that route to look through https://code.videolan.org/search?utf8=%E2%9C%93&search=AddStaticLink.sh&group_id=9&project_id=435&scope=&search_code=true&snippets=false&repository_ref=master&nav_source=navbar and make sure you know which ones actually need it

**Comment by revelator** [comment]
wonder if this script could be added to qt5-static to fixup the package config files ?.

**Comment by davidgiven** [comment]
+1 to that last --- I think I've just run into the same problem...
